GLMakie/src/drawing_primitives.jl

```diff
@@ -47,47 +47,36 @@ end
 make_context_current(screen::Screen) = GLFW.MakeContextCurrent(to_native(screen))
 
 function connect_camera!(gl_attributes, cam, space = gl_attributes[:space])
-    # (0, 1) x (0, 1) x (0, 1) space
-    rel = Mat4f(2, 0, 0, 0, 0, 2, 0, 0, 0, 0, 1, 0, -1, -1, 0, 1)
-    # (-1, 1) x (-1, 1) x (0, 1) space
-    id = Mat4f(I)
-
-    for key in (:pixel_space, :view, :projection, :resolution, :eyeposition, :projectionview)
-        if !haskey(gl_attributes, key)
-            gl_attributes[key] = if key in (:projection, :projectionview)
-                map(getfield(cam, key), getfield(cam, :pixel_space), space) do data, pixel, space
-                    if is_data_space(space)
-                        return data
-                    elseif is_pixel_space(space)
-                        return pixel
-                    elseif is_relative_space(space)
-                        return rel
-                    elseif is_clip_space(space)
-                        return id
-                    else
-                        error("Space $space not recognized. Must be one of $(spaces())")
-                    end
-                end
-            elseif key in (:view, )
-                map(getfield(cam, key), space) do view, space
-                    ifelse(is_data_space(space), view, id)
-                end
-            else # :pixel_space, :resolution, :eyepostion
-                getfield(cam, key)
-            end
+    for key in (:pixel_space, :resolution, :eyeposition)
+        get!(gl_attributes, key, getfield(cam, key))
+    end
+    get!(gl_attributes, :view) do
+        map(cam.view, space) do view, space
+            return is_data_space(space) ? view : Mat4f(I)
         end
     end
-
-    if !haskey(gl_attributes, :normalmatrix)
-        gl_attributes[:normalmatrix] = map(gl_attributes[:view], gl_attributes[:model]) do v, m
+    get!(gl_attributes, :normalmatrix) do
+        return map(gl_attributes[:view], gl_attributes[:model]) do v, m
             i = Vec(1, 2, 3)
             return transpose(inv(v[i, i] * m[i, i]))
         end
     end
 
+    get!(gl_attributes, :projection) do
+        Makie.space_to_clip_obs(cam, space[], false)
+        return map(cam.projection, space) do _, space
+            return Makie.space_to_clip(cam, space, true)
+        end
+    end
```

Comment:
The matrix here can be `cam.projection` or `cam.pixel_space` so it should trigger on both of those and `space`. Also seems like you didn't update this correctly

Reply:
I think for 2d, 3d and pixel cameras reacting to `pixelspace` isn't necessary but for `RelativeCamera` it would be?

Reply:
hm, for simplicity I'll just add pixel_space to it for now...
The right way to do this is hard to express with observables... I need to think about a cleaner way going forward...